lib/grp.gi

```diff
@@ -783,16 +805,35 @@ InstallMethod( DerivedSeriesOfGroup,
     D := DerivedSubgroup( G );
    
     while
-      # we don't know that the group has no generators
-      (not HasGeneratorsOfGroup(S[Length(S)]) or
-	    Length(GeneratorsOfGroup(S[Length(S)]))>0) and
-      ( (not HasAbelianInvariants(S[Length(S)]) and D <> S[ Length(S) ]) or
-	    Length(AbelianInvariants(S[Length(S)]))>0) do
+      (not HasIsTrivial(S[Length(S)]) or
+	    not IsTrivial(S[Length(S)])) and
+      (
+        (not HasIsPerfectGroup(S[Length(S)]) and
+         not HasAbelianInvariants(S[Length(S)]) and D <> S[ Length(S) ]) or
+        (HasIsPerfectGroup(S[Length(S)]) and not IsPerfectGroup(S[Length(S)]))
+        or (HasAbelianInvariants(S[Length(S)])
+                            and Length(AbelianInvariants(S[Length(S)])) > 0)
+      ) do
```

Comment:
@fingolfin 
Should not ````IsAttributeKnownAndEqualTo````, and similarly ````IsKnownToBeTrivial````, go to some other file, like oper1.gi? Anyway, this seems like more than what this PR is for. I do not mind to wait with this PR until such functions are available, but such a useful function would be lost in this PR which is about something else. 

BTW, your code really is not entirely the same as the original, but I get the idea. In any case, the code runs fine for e.g. ````Group((1,2,3),(3,4,5))````, and the key is the ````D <> S[ Length(S) ])```` part of the code.